Implement cbuild-pack.yml support

Torbjorn-Svensson opened the pull request:
Fixes #1122

Contributed by STMicroelectronics

brondani commented:
@Torbjorn-Svensson @ErikMallbergSTM Can you please explain the algorithm being implemented here? I see some pack resolution stuff that I wouldn't expect. My expectation would be to store in `cbuild-pack.yml` the information about processed packs similarly to what is already collected and stored in `cbuild.yml` files.
I don't care about internal implementation specific names, but externally in yml files and schemas the naming convention must be consistent, if possible please do not mix terms such as `cbuild-pack` and `lockfile`.

Torbjorn-Svensson commented:
> @Torbjorn-Svensson @ErikMallbergSTM Can you please explain the algorithm being implemented here? I see some pack resolution stuff that I wouldn't expect. My expectation would be to store in `cbuild-pack.yml` the information about processed packs similarly to what is already collected and stored in `cbuild.yml` files. I don't care about internal implementation specific names, but externally in yml files and schemas the naming convention must be consistent, if possible please do not mix terms such as `cbuild-pack` and `lockfile`.

Thanks for your input.
We have tried to address your comments in the last update.
Documentation of the algorithm is on the way and will be shared soon.

brondani commented on tools/projmgr/src/ProjMgrYamlEmitter.cpp:
```diff
@@ -71,6 +71,121 @@ class ProjMgrYamlCbuildIdx : public ProjMgrYamlBase {
+      const string packId = RtePackage::ComposePackageID(reqInfo.vendor, reqInfo.name, reqInfo.version);
+      if (model.find(packId) == model.end()) {
+        // Pack does not have a version range, so add it
+        ModelItem modelItem;
+        ProjMgrUtils::ConvertToPackInfo(packId, modelItem.info);
+        modelItem.resolvedPack.pack = packId;
+        modelItem.resolvedPack.selectedBy.push_back(packId);
+        model[packId] = modelItem;
+      }
+    }
+  }
+
+  // Stage 4: Add all original expressions that was eagerly replaced
+  for (const auto& context : processedContexts) {
+    for (const auto& [key, val] : context->eagerlyReplacedPackIdMap) {
+      if (model.find(val) != model.end()) {
+        ProjMgrUtils::PushBackUniquely(model[val].resolvedPack.selectedBy, key);
+      }
+    }
+  }
+
+  // Stage 5: Process all wildcard patterns from user and add to selected-by list
+  for (const auto& context : processedContexts) {
+    for (const auto& packItem : context->packRequirements) {
+      // Skip project local packs
+      if (!packItem.path.empty()) {
+        continue;
+      }
+
+      const PackInfo& reqInfo = packItem.pack;
+      if (reqInfo.name.empty() || WildCards::IsWildcardPattern(reqInfo.name)) {
+        const string packId = RtePackage::ComposePackageID(reqInfo.vendor, reqInfo.name, reqInfo.version);
+
+        for (auto& [key, val] : model) {
+          if (ProjMgrUtils::IsMatchingPackInfo(val.info, reqInfo)) {
+            ProjMgrUtils::PushBackUniquely(val.resolvedPack.selectedBy, packId);
+          }
+        }
+      }
+    }
```
For consistency and avoiding duplication, if possible I would recommend keeping the pack resolution in a common place, currently I think it’s done mostly in `ProjMgrWorker::GetRequiredPdscFiles`, and to store the information needed by the `selected-by` node close to the relevant variables in the pack resolution chain (`context.packRequirements`, `context.pdscFiles`, `m_loadedPacks`), eventually simplifying the emitter code here and also covering local packs that are commented as TODO.

Torbjorn-Svensson replied:
Project local packs are not included in the `cbuild-pack.yml` file.

The loading and the persisting of the `cbuild-pack.yml` content needs to be done before respectively after the RTE model has decided what packs that are to be used. See the diagram in the [documentation](https://github.com/slhultgren/cmsis-toolbox/blob/pr/shu-cbuild-pack-documentation/docs/YML-CBuild-Format.md#pack-locking) for details.

brondani commented on tools/projmgr/src/ProjMgrYamlEmitter.cpp:
```diff
@@ -596,6 +711,9 @@ bool ProjMgrYamlBase::WriteFile(YAML::Node& rootNode, const std::string& filenam
       ProjMgrLogger::Error(filename, "file cannot be written");
       return false;
     }
+    if (!schemaName.empty()) {
+      fileStream << "# yaml-language-server: $schema=https://raw.githubusercontent.com/Open-CMSIS-Pack/devtools/main/tools/projmgr/schemas/" << schemaName << ".schema.json" << endl << endl;
+    }
```
The `# yaml-language-server` comment has higher priority over IDE workspace and user selections. To avoid hiding such selections it's recommended not to print this comment by default.

Torbjorn-Svensson commented:
We are working on improving the documentation and addressing the remaining comments pr and in https://github.com/Open-CMSIS-Pack/cmsis-toolbox/pull/76.

Torbjorn-Svensson commented:
Hi, @soumeh01 were you suppose to do merges into this branch or some other branch? Otherwise, we will just rebase when we do the next push since there are still a few changes left.

soumeh01 commented:
> Hi, @soumeh01 were you suppose to do merges into this branch or some other branch? Otherwise, we will just rebase when we do the next push since there are still a few changes left.

Ok, I was not aware there were still changes left and had updated the branch.
No worries we can wait for the leftover changes and then you can rebase it with main.

Torbjorn-Svensson commented:
Some tests remain to be created to verify usage of context selection.

Torbjorn-Svensson commented:
@jkrech @brondani This PR is now updated with the additional test. Please review it and see if there is anything left before this PR can be merged.

Torbjorn-Svensson commented:
@jkrech Can you fix the coverage job that failed for some reason?

jkrech commented:
I triggered the coverage action again and it completed successfully now.

brondani commented on tools/projmgr/src/ProjMgrYamlEmitter.cpp:
```diff
@@ -71,6 +71,143 @@ class ProjMgrYamlCbuildIdx : public ProjMgrYamlBase {
+    if (packInfo1.vendor == packInfo2.vendor && packInfo1.name == packInfo2.name) {
+      return VersionCmp::Compare(packInfo1.version, packInfo2.version) < 0;
+    }
+    return packInfo1.vendor < packInfo2.vendor || packInfo1.name < packInfo2.name;
```
Depending on input data this can lead to `invalid comparator` failing, for example when sorting `ARM::CMSIS` and `Keil::B-U585I-IOT02A_BSP` packs.
From the coverage report it seems there is no test case for this branch:
https://app.codecov.io/gh/Open-CMSIS-Pack/devtools/pull/1143/blob/tools/projmgr/src/ProjMgrYamlEmitter.cpp#L191

brondani commented on tools/projmgr/schemas/cbuild-pack.schema.json:
```diff
@@ -0,0 +1,13 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "$id": "https://raw.githubusercontent.com/Open-CMSIS-Pack/devtools/schemas/projmgr/2.1.0/tools/projmgr/schemas/cbuild-pack.schema.json",
+  "title": "CMSIS cbuild-pack",
+  "description": "file containing all pack versions used, used for pack locking",
+  "version": "2.1.0",
+  "properties": {
+    "cbuild-pack": {
+      "$ref": "./common.schema.json#/definitions/BuildPackDescType"
+    }
+  },
+  "required": [ "cbuild-pack" ]
+}
```
As recently requested in https://github.com/Open-CMSIS-Pack/devtools/issues/1211 please add the type definition (e.g. `"type": "object"`) and possibly the recommended `"additionalProperties": false`.

Torbjorn-Svensson commented:
@jkrech Looks like the coverage failed again. Can someone at Arm look into why this job appears to fail repeatedly?
For reference: https://github.com/Open-CMSIS-Pack/devtools/actions/runs/6942633876/job/18885984864?pr=1143

soumeh01 commented:
> @jkrech Looks like the coverage failed again. Can someone at Arm look into why this job appears to fail repeatedly? For reference: https://github.com/Open-CMSIS-Pack/devtools/actions/runs/6942633876/job/18885984864?pr=1143


@Torbjorn-Svensson, it appears there's a Codecov server outage. Despite 3 attempts to upload coverage data, the job has failed. We'll need to wait for all the jobs to complete before I can retrigger the coverage job.

Torbjorn-Svensson commented:
> > @jkrech Looks like the coverage failed again. Can someone at Arm look into why this job appears to fail repeatedly? For reference: https://github.com/Open-CMSIS-Pack/devtools/actions/runs/6942633876/job/18885984864?pr=1143
> 
> @Torbjorn-Svensson, it appears there's a Codecov server outage. Despite 3 attempts to upload coverage data, the job has failed. We'll need to wait for all the jobs to complete before I can retrigger the coverage job. Issue: [codecov/codecov-action#598](https://github.com/codecov/codecov-action/issues/598)

As the same problem have happened a few times in a row now, would it be possible to give me the power to be able to retrigger the failed jobs?

jkrech commented:
@Torbjorn-Svensson can you try to see whether you can see the re-run action now?

brondani commented on test/packs/SomeVendor/RteTest/0.0.1/SomeVendor.RteTest.pdsc:
Please note that adding test data into `test` folder may require other tools/libs tests to be updated.
I see RteModelTest is failing:
```
[ RUN      ] RteModelTest.LoadPacks
libs\rtemodel\test\src\RteModelTest.cpp(59): error: Expected equality of these values:
  latestFiles.size()
    Which is: 7
  6
libs\rtemodel\test\src\RteModelTest.cpp(63): error: Expected equality of these values:
  files.size()
    Which is: 8
  7
```

Torbjorn-Svensson commented on tools/projmgr/schemas/common.schema.json:
```diff
@@ -931,6 +931,32 @@
+        }
+      },
+      "additionalProperties": false,
+      "required": [ "resolved-pack" ]
```
Yes, in the case where there is no `packs` node in the yml files. Then we can get either though device or component usage, but there is nothing explicitly specifying it in the user provided lists.

Torbjorn-Svensson commented on tools/projmgr/test/src/ProjMgrUnitTests.cpp:
```diff
@@ -696,6 +700,369 @@ TEST_F(ProjMgrUnitTests, RunProjMgrLayers) {
+  const string cbuildPack = testinput_folder + "/TestSolution/PackLocking/project_pack_lock_with_for_context.cbuild-pack.yml";
+  const string output = testoutput_folder + "/testpacklock";
+
+  // Ensure clean state for the test case.
```
Should be tested in `ProjMgrUnitTests.RunProjMgrSolution_CbuildPackWithoutUsedComponents`.

Torbjorn-Svensson commented:
> @Torbjorn-Svensson can you try to see whether you can see the re-run action now?

I think I have what's needed now. I'll verify next time I see a broken build if I can retrigger.
